src/internal/connector/graph/middleware.go

```diff
@@ -419,13 +424,55 @@ func extractRateLimiterConfig(ctx context.Context) (LimiterCfg, bool) {
 	return lc, ok
 }
 
+type limiterConsumptionKey string
+
+const (
+	limiterConsumptionCtxKey       limiterConsumptionKey = "corsoGraphRateLimiterConsumption"
+	defaultLimiterConsumption                            = 1
+	driveDefaultLimiterConsumption                       = 2
+	// limit consumption rate for single-item GETs requests,
+	// or delta-based multi-item GETs.
+	SingleGetOrDeltaLC = 1
+	// limit consumption rate for anything permissions related
+	PermissionsLC = 5
+)
+
+// ConsumeNTokens ensures any calls using this context will consume
+// n rate-limiter tokens.  Default is 1, and this value does not need
+// to be established in the context to consume the default tokens.
+// This should only get used on a per-call basis, to avoid cross-pollination.
+func ConsumeNTokens(ctx context.Context, n int) context.Context {
+	return context.WithValue(ctx, limiterConsumptionCtxKey, n)
+}
+
+func ctxLimiterConsumption(ctx context.Context, defaultConsumption int) int {
+	l := ctx.Value(limiterConsumptionCtxKey)
+	if l == nil {
+		return defaultConsumption
+	}
+
+	lc, ok := l.(int)
+	if !ok || lc < 1 {
```

Comment:
Would be good to add an upper bound check too.

Reply:
Upper bounds are naturally checked by the rate limiter.  If a capacity greater than the per-second burst is requested, then the limiter (and thus api call) will error.  Should be clearly identifiable in CI testing, and prevents us from needing to check against N different upper bounds (one for each limiter implementation, should they differ).